### PR TITLE
Add find-includes script

### PR DESCRIPTION
**Pull request by richsalz**
This is a "good enough" script to quickly find included headers. Runs in seconds.  Not ported to Windows or VMS.

I think this will also be useful for generating the FIPS checkums to have *just* the header files.

Needs work, hence a draft, but would like feedback from @levitte, @mspncp , @t8m who have commented on these issues already.

### CODE REVIEW

**Comment by paulidale** [comment]
Might the `-MM` option to GNU's cpp circumvent a lot of the need for this?

**Comment by richsalz** [comment]
I didn't look, but I thought we were doing something like `-MM`  We certainly should be.  This is an attempt to do the equivalent on other platforms. And then it also became useful, IMO, for the FIPS checksums with a couple of lines of code. Requiring the FIPS checksums to be built only on linux seems needlessly limiting to other people who want to build their own derived module (that they are getting validated) on their own platform/tools.

**Comment by richsalz** [comment]
@t8m, @levitte is this useful for adding the headers to the checksum? Is anything missing?
Also, is there interest in using this on Windows and/or VMS? I think it's much faster, @mspncp.

If it's not of interest to either use, please let me know and I will close it.

**Comment by paulidale** [comment]
I rather suspect that #15365 is a more robust approach to this.  It's a lot slower.

File: util/find-includes.pl [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+    open(my $in, "<", $f)
+        or die "Can't open $f, $!, ";
+
+    my $dirty = 0;
+    while ( <$in> ) {
+        next unless /^# *include */;
+        if ( m/include *"([^"]*)"/ ) {
+            printit($f, $1);
+            $dirty = 1;
+        } elsif ( /include *<([^>]*)>/ ) {
```
[the remark this answers is not in the document]
Comment (richsalz) [reply util/find-includes.pl:60]: cute hack.  I did this:
```
g grep 'include <' | sed -e /openssl/d -e /crypto/d -e /internal/d \
-e /string.h/d -e/stdio.h/d -e/ctype.h/d -e/errno.h/d -e /assert.h/d

```
and skimmed the list and didn't find anything.  I'll do that and push a commit shortly.

File: util/find-includes.pl [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    } else {
+        die "Unknown flag $arg, ";
+    }
+}
+
+# Hacks.
+$headers{'e_os.h'} = 1;
+
+# Skip obvious system headers.
+foreach my $s ( glob("/usr/include/*.h /usr/include/sys/*.h") ) {
```
[the remark this answers is not in the document]
Comment (richsalz) [reply util/find-includes.pl:103]: That's okay.  It's just an optimization in the `printit` sub.

File: util/find-includes.pl (hunk 1 above)
[the remark this answers is not in the document]
Comment (paulidale) [reply util/find-includes.pl:60]: Isn't this deleting the includes with openssl, crypto and internal?

**Comment by richsalz** [comment]
> I rather suspect that #15365 is a more robust approach to this. It's a lot slower.

Wanna run a diff to see if this is missing things?  I will in a few hours if I have time.

File: util/find-includes.pl (hunk 1 above)
[the remark this answers is not in the document]
Comment (richsalz) [reply util/find-includes.pl:61]: Isn't this deleting the includes with openssl, crypto and internal?

Yes.  I wanted to see if anything was being missed if we checked for just those three directories.
And then the second line filtered out some of the standard C noise.

**Comment by richsalz** [comment]
So the question now becomes, is this worth doing for non-gcc platforms to speed up their build times?
Ping @levitte to answer about VMS, and I don't know who knows anything about Windows.

**Comment by richsalz** [comment]
Here I am, trying to help platforms I don't support (windows and vms) and make their builds faster and it gets ignored :)

Closing for lack of interest.

**Comment by t8m** [comment]
@richsalz Actually this might be a way how to fix #15639. I think this can still be considered and used in that context after the beta1 release.

**Comment by levitte** [comment]
We do use `-MM` (in fact, `-MMD`) with gcc like compilers. A very simple grep in Makefile on most Linux boxes would have answered that.

We do the same thing on VMS already, and we have a similar albeit a bit clumsy variant on Windows too.

What's needed to find the appropriate header files for the FIPS module is to find all the object files that went into it, and read the corresponding `.d` files. They have the dependency info.

**Comment by t8m** [comment]
So in the end it was not necessary to use the script to fix #15639. So this can be closed IMO.

**Pull request closed without merging**